### ISSUE TITLE
Discussion: Creating area-specific weights for a large number of areas

### ISSUE THREAD

**Issue by donboyd5** [PSLmodels/tax-microdata-benchmarking#197]
It appears increasingly likely that it would be useful to produce area-specific weights for a large number of areas.

That raises at least 3 questions to think about:

1. Automating creation of multiple sets of weights
2. Pre-optimization identification of potential problems and possible pre-optimization adjustments
3. Post-optimization diagnosis of actual problems

Here are some thoughts on each.

**Automating creation of multiple sets of weights**

Because IRS data for all Congressional Districts will be formatted similarly and all states will be formatted similarly, I expect that I will create one big file for each category, with an extra field with a code for the CD or state. I could even combine the two files into one, with the code determining whether it is a state or a CDIf we keep the current setup for areas, I'll just extract individual CDs and states into their own files, as needed (or perhaps create all files at once).

However, it might also make sense to set the program up to do many areas sequentially, driven by a file that has multiple areas (possibly a subset of the big file I just described, with areas and targets tailored to the problem at hand). In this case we'd want to have some sort of "apply" ability to apply a create-area-file function mutiple times.

This would create some new considerations because doing many areas and bulk will mean that we probably will not have the time (human time - yours and mine, not computer time) to give loving attention to each area and its targets.

**Pre-optimization identification of potential problems and possible pre-optimization adjustments**

If we do areas in bulk, we'll want some way to identify potential problems and possibly even adjust for them in an automated way.

Identification probably entails identifying targets for individual areas that appear hard, and areas that appear to have combinations of targets that are hard. At least intially, we'll want to use a fairly simple approach I think.

**Post-optimization diagnosis of actual problems**

**Comment by martinholmer** [comment]
@donboyd5 said in issue #197:

> However, it might also make sense to set the program up to do many areas sequentially, driven by a file that has multiple areas (possibly a subset of the big file I just described, with areas and targets tailored to the problem at hand). In this case we'd want to have some sort of "apply" ability to apply a create-area-file function mutiple times.

Both of these approaches seem to be at variance with the approach we decided on weeks ago and for which `create_areas_weights.py` is designed to handle.   Each area needs a `{area}_targets.csv` file in the `areas/targets` folder and `create_areas_weights.py` will produce a `{area}_tmd_weights.csv.gz` file in the `areas/weights` folder.   We can discuss how to automate this so weight files are not unnecessarily produced (sort of like a Makefile works).

Here is the docstring at the top of `create_areas_weights.py`:
```
"""
Construct AREA_tmd_weights.csv.gz, a Tax-Calculator-style weights file
for 2021+ for the specified sub-national AREA.

AREA prefix for state areas are the two lower-case character postal codes.
AREA prefix for congressional districts are the state prefix followed by
two digits (with a leading zero) identifying the district.  There are no
district files for states with only one congressional district.
"""
```

**Comment by donboyd5** [comment]
That's true. It seems valuable to me, especially in light of Matt's interest in hitting a wide range of CDs . But perhaps the value does not outweigh the work involved. I understand completely the need to weigh value against work and am not trying to force it.
